src/os/bluestore/BlueStore.cc

```diff
@@ -698,48 +702,103 @@ void BlueStore::TwoQCache::trim(uint64_t onode_max, uint64_t buffer_max)
-      Buffer *b = &*buffer_hot.rbegin();
-      assert(b->is_clean());
-      dout(20) << __func__ << " buffer_hot rm " << *b << dendl;
-      b->space->_rm_buffer(b);
-    }
-    while (buffer_warm_in.size() > kin) {
-      Buffer *b = &*buffer_warm_in.rbegin();
+  // first adjust clean(warm_in and hot) buffer lists
+  if (buffer_bytes > buffer_max) {
+
+    int64_t to_evict_bytes = buffer_bytes - buffer_max;
+    uint64_t reclaimed = 0;
+
+    // evict warm_in foremost
+    while (to_evict_bytes > 0) {
+      auto p = buffer_warm_in.rbegin();
+      if (p == buffer_warm_in.rend()) {
+        // stop if warm_in list is now empty
+        break;
+      }
+
+      Buffer *b = &*p;
       assert(b->is_clean());
       dout(20) << __func__ << " buffer_warm_in -> out " << *b << dendl;
       buffer_bytes -= b->length;
+      buffer_list_bytes[BUFFER_WARM_IN] -= b->length;
+      to_evict_bytes -= b->length;
+      reclaimed += b->length;
       b->state = Buffer::STATE_EMPTY;
       b->data.clear();
       buffer_warm_in.erase(buffer_warm_in.iterator_to(*b));
       buffer_warm_out.push_front(*b);
       b->cache_private = BUFFER_WARM_OUT;
+      buffer_list_bytes[BUFFER_WARM_OUT] += b->length;
     }
-    while (buffer_warm_out.size() > kout) {
-      Buffer *b = &*buffer_warm_out.rbegin();
-      assert(b->is_empty());
-      dout(20) << __func__ << " buffer_warm_out rm " << *b << dendl;
+
+    if (to_evict_bytes > 0) {
+      dout(20) << __func__ << " reclaimed " << prettybyte_t(reclaimed)
+               << " from warm_in list, still have "
+               << prettybyte_t(to_evict_bytes) << "to reclaim"
+               << dendl;
+    } else {
+      dout(20) << __func__ << " reclaimed " << prettybyte_t(reclaimed)
+               << " from warm_in list, done reclaiming clean buffers"
+               << dendl;
+    }
+
+    reclaimed = 0;
+    // then evict hot
+    while (to_evict_bytes > 0) {
+      auto p = buffer_hot.rbegin();
+      if (p == buffer_hot.rend()) {
+        // stop if hot list is now empty
+        break;
+      }
+
+      Buffer *b = &*p;
+      assert(b->is_clean());
+      dout(20) << __func__ << " buffer_hot rm " << *b << dendl;
+      // adjust evict size before buffer goes invalid
+      to_evict_bytes -= b->length;
+      reclaimed += b->length;
       b->space->_rm_buffer(b);
     }
+
+    if (to_evict_bytes > 0) {
+      // actually this should not happen, but tolerate it for now
+      dout(20) << __func__ << " reclaimed " << prettybyte_t(reclaimed)
+               << " from hot list, still have "
+               << prettybyte_t(to_evict_bytes) << "to reclaim"
+               << dendl;
+    } else {
+      dout(20) << __func__ << " reclaimed " << prettybyte_t(reclaimed)
+               << " from hot list, done reclaiming clean buffers"
+               << dendl;
+    }
+  }
+
+  // adjust ghost(warm_out) list too, if necessary
+  uint64_t& ghost_list_bytes = buffer_list_bytes[BUFFER_WARM_OUT];
```

Comment:
also, i'm not sure it makes sense to count warm_out in terms of bytes.  the previous behavior of using average buffer size to count buffers seems like it would be more fair?